tests/Unit/NumericalAlgorithms/Interpolation/Test_CubicSpline.cpp

```diff
@@ -0,0 +1,113 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <cmath>
+#include <cstddef>
+#include <random>
+#include <vector>
+
+#include "NumericalAlgorithms/Interpolation/CubicSpline.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace {
+template <class F>
+void test_cubic_spline(const F& function, const double lower_bound,
+                       const double upper_bound, const size_t size,
+                       const double tolerance,
+                       const double tolerance_interior) noexcept {
+  // Construct random points between lower and upper bound to interpolate
+  // through. Always include the bounds in the x-values.
+  std::vector<double> x_values(size), y_values(size);
+  const double delta_x = (upper_bound - lower_bound) / size;
+  MAKE_GENERATOR(gen);
+  std::uniform_real_distribution<> dist(0., delta_x);
+  x_values.front() = lower_bound;
+  for (size_t i = 1; i < size - 1; ++i) {
+    x_values[i] = lower_bound + i * delta_x + dist(gen);
+  }
+  x_values.back() = upper_bound;
+  for (size_t i = 0; i < size; ++i) {
+    y_values[i] = function(x_values[i]);
+  }
+
+  Approx custom_approx = Approx::custom().epsilon(tolerance).scale(1.);
+  Approx custom_approx_interior =
+      Approx::custom().epsilon(tolerance_interior).scale(1.);
+
+  // Construct the interpolant and give an example
+  /// [interpolate_example]
+  intrp::CubicSpline interpolant{x_values, y_values};
+  const double x_to_interpolate_to = (upper_bound - lower_bound) / 2.;
+  CHECK(interpolant(x_to_interpolate_to) ==
+        custom_approx_interior(function(x_to_interpolate_to)));
+  /// [interpolate_example]
+
+  // Check that the interpolation is exact at the datapoints
+  for (const auto& x_value : x_values) {
+    CHECK(interpolant(x_value) == approx(function(x_value)));
+  }
+
+  // Check that the interpolation matches the function within the given
+  // tolerance. Also check that the serialized-and-deserialized interpolant does
+  // the same.
+  const auto deserialized_interpolant = serialize_and_deserialize(interpolant);
+  double max_error = 0.;
+  double max_error_interior = 0.;
+  for (size_t i = 0; i < 10 * size; ++i) {
+    const double x_value = lower_bound + i * delta_x * 0.1 + 0.1 * dist(gen);
+    CAPTURE(x_value);
+    const double y_value = function(x_value);
+    const double interpolated_y_value = interpolant(x_value);
+    max_error = std::max(abs(interpolated_y_value - y_value), max_error);
+    CHECK(interpolated_y_value == custom_approx(y_value));
+    CHECK(deserialized_interpolant(x_value) == approx(interpolated_y_value));
```

Comment:
This should be exact.

Reply:
Not sure if the `gsl_intrp_acc` accelerator messes that up, since I think that‘s some kind of stateful / caching mechanism. I‘ll try.

Reply:
seems fine